### DISCUSSION TITLE
Type-safe middleware `composition`-ing

### DISCUSSION THREAD

**Discussion by iam-kevin** [trpc/trpc#3890] [RFC / Ideas]
Hey there. Great fan of your work and love trpc.

Wanted to know if there could be a feature that allowed inferred middleware composition.
The composed function should be able to defined what types it expects from the composing function.

Consider the following scenario when you are have a school system where you have either a `student`, `teacher` and `bot` entity. You then want to create reusable procedures that can be type infered.
 
```ts
import * as t from "~/trpc";

/**
 * Middleware that extracts that information of the user as a teacher
 */
const asTeacherMiddleware = t.middleware(async ({ ctx, next }) => {
	return next({
		ctx: {
			...ctx,
			teacher: { id: "tch_id", name: "Jane Doe" },
			canPost: false,
		},
	});
});

// extracting user information to use as student
const asStudentMiddleware = t.middleware(async ({ ctx, next }) => {
	return next({
		ctx: {
			...ctx,
			student: { id: "sdt_abc", name: "John Smith" },
			canPost: true,
		},
	});
});

// nothing is changed
const asNothingMiddleware = t.middleware(async ({ next }) => next());
```

Using a `extendContext` function to tell middleware that that its should expect the function it's composed with will have certain other properties.
```ts
const isAllowedToPost = t.extendContext<{ canPost: boolean }>().middleware(async ({ ctx, next }) => {
	if (!ctx.canPost) {
		// ^? `.canPost` is expected to be present in ctx (not undefined)
		throw new t.TRPCError({ code: "FORBIDDEN" });
	}

        ...
	// proceeds if allowed
	return next({ ctx });
});
```

When being composed as a procedure, it should be type-safe like this

```ts
const teacherPosterProcedure = t.procedure
	.use(asTeacherMiddleware)
	.use(isAllowedToPost);
const studentPosterProcedure = t.procedure
	.use(asTeacherMiddleware)
	.use(isAllowedToPost);

// This should't work (on a typescript level)
const thisShouldntWorkProcedure = t.procedure
	.use(asNothingMiddleware)
	.use(isAllowedToPost); // Type-Error: Expected { canPost: boolean } as part of context
        ~~~~~~~~~~~~~~~~
```

**Comment by iam-kevin** [comment]
Current work around is adding `{ canPost?: true }` as part of `createContext`:

```ts
export async function createContext({
 req,
}: trpcExpress.CreateExpressContextOptions) {
 return {
   query: req.query,
 } as {
   canPost?: boolean,
   authorization?: string
  }
}
```

But that section gets bloated pretty quickly when adding multiple partial information for other features you'd want.

```ts

export async function createContext({
 req,
}: trpcExpress.CreateExpressContextOptions) {
 return {
   query: req.query,
 } as {
   canPost?: boolean,
   authorization?: string,
   scope?: string[],
   teacher?: { id: string, name: string },
   student?: { id: string, name: string },
   bot?: { id: string, key: string  },
   ...
  }
}
 ```

**Comment by KATT** [comment]
Have you seen https://trpc.io/docs/middlewares#extending-middlewares?

**Reply by iam-kevin** [reply]
Yes.

Ideal for when you want the updated context to trickle down. 

i.e. 
```ts
// current scenario
middleware_a.pipe(middleware_b) 
```
here, context modified in `middleware_a` is now part of context in `middleware_b`


But can't seem to get it to work when I want the appended middleware (`middleware_b`) to expect a certain values from it's piping middleware `middleware_a`

**What I'm hoping to be able to do...**
```ts
// something like
let middleware_b: MiddlewareFunction<{ foo: "bar" }> = t.middleware(...)

// then 
middleware_a.pipe(middleware_b) 
```
`middleware_b` expects that the context modified by `middleware_a` to have `{..., foo: "bar", ...}`
